source/particle/property/cpo.cc

```diff
@@ -0,0 +1,1124 @@
+/*
+  Copyright (C) 2020 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file LICENSE.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/property/cpo.h>
+#include <world_builder/grains.h>
+#include <aspect/geometry_model/interface.h>
+#include <world_builder/world.h>
+
+#include <aspect/utilities.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Property
+    {
+
+      template <int dim>
```

Comment:
The are static variables, so they need to be declared in the source file.

Reply:
I wonder whether they need to be static, but I suppose it makes sense that you would want a single class-wide value.

Reply:
part of the reason why they are static is because it is information I would like to directly use in particle property plugins and static particle property functions (such as the store and load functions, which are use by these other plugins) which directly builds on this information.